run/pytorch-train.py

```diff
@@ -0,0 +1,160 @@
+            "coiled",
+            "pytorch",
+            "cudatoolkit",
+            "pynvml",
+            "torchvision",
+        ],
+    },
+    gpu_enabled=True,
+)
+
+
+def load_data():
+    transform = transforms.Compose(
+        [transforms.ToTensor(),
+        transforms.Normalize((0.5,), (0.5,))])
+
+    # Create datasets for training & validation, download if necessary
+    training_set = datasets.FashionMNIST(os.getcwd(), train=True, transform=transform, download=True)
+    validation_set = datasets.FashionMNIST(os.getcwd(), train=False, transform=transform, download=True)
+
+    # Create data loaders for our datasets; shuffle for training, not for validation
+    training_loader = torch.utils.data.DataLoader(training_set, batch_size=4, shuffle=True)
+    validation_loader = torch.utils.data.DataLoader(validation_set, batch_size=4, shuffle=False)
+
+    # Report split sizes
+    print('Training set has {} instances'.format(len(training_set)))
+    print('Validation set has {} instances'.format(len(validation_set)))
+
+    return training_loader, validation_loader
+
+
+class GarmentClassifier(nn.Module):
+    def __init__(self):
+        super(GarmentClassifier, self).__init__()
+        self.conv1 = nn.Conv2d(1, 6, 5)
+        self.pool = nn.MaxPool2d(2, 2)
+        self.conv2 = nn.Conv2d(6, 16, 5)
+        self.fc1 = nn.Linear(16 * 4 * 4, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84, 10)
+
+    def forward(self, x):
+        x = self.pool(F.relu(self.conv1(x)))
+        x = self.pool(F.relu(self.conv2(x)))
+        x = x.view(-1, 16 * 4 * 4)
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+def train_one_epoch(model, loss_fn, optimizer, training_loader):
+    running_loss = 0.
+    last_loss = 0.
+
+    # Here, we use enumerate(training_loader) instead of
+    # iter(training_loader) so that we can track the batch
+    # index and do some intra-epoch reporting
+    for i, data in enumerate(training_loader):
+        # Every data instance is an input + label pair
+        inputs, labels = data
+
+        # Zero your gradients for every batch!
+        optimizer.zero_grad()
+
+        # Make predictions for this batch
+        outputs = model(inputs)
+
+        # Compute the loss and its gradients
+        loss = loss_fn(outputs, labels)
+        loss.backward()
+
+        # Adjust learning weights
+        optimizer.step()
+
+        # Gather data
+        running_loss += loss.item()
+        if i % 1000 == 999:
+            last_loss = running_loss / 1000 # loss per batch
+            print('  batch {} loss: {}'.format(i + 1, last_loss))
+            running_loss = 0.
+
+    return last_loss
+
+@coiled.run(
+    vm_type="g5.xlarge",  # A GPU Instance Type
+    software="pytorch",   # Our software environment defined above
+    region="us-west-2",   # We find GPUs are easier to get here
+)
+def train_all_epochs():
```

Comment:
I would have expected somewhere in here a PyTorch line specifying that we're using the cuda backend.  My guess is that we're just using CPU here.  If you search the previous example for `cuda` you'll find the relevant line.

Reply:
I pushed https://github.com/coiled/examples/pull/22/commits/f5df826f00f613b94d0e056e86889220d98ebc07 with code changes so that model trains on GPU. @scharlottej13 if you want to revert that / do something different / make changes, go ahead (and hope you don't mind me jumping in, I wanted to do something fun this morning).

You can see the logging I added for the cluster I ran: https://staging.coiledhq.com/clusters/93357/information?account=nat-tabris-staging&tab=Logs&filterPattern=instance%3Ascheduler+cuda

Reply:
Thanks @ntabris!

Reply:
made a small change, looking good https://cloud.coiled.io/clusters/245450/information?account=sarah-johnson&tab=Metrics

Reply:
I'm running this now and GPU util is pretty sad (but also non-zero), pretty much sitting at 13–14%.

Reply:
which cluster? I was wondering if this example would hold up comparing to CPU-only (b/c of the cost of loading data to the GPU) https://cloud.coiled.io/clusters/245460/information?account=sarah-johnson&tab=Metrics but hadn't looked too deeply at GPU util yet

Reply:
nvm found it https://cloud.coiled.io/clusters/245461/information?account=nat-tabris&tab=Metrics

Reply:
I'm not seeing any GPUs w/ less than 4 vCPU, maybe we could do some heavier computation in the model training?

Reply:
I think low CPU util is fine here, I was noting that GPU util was low. I actually wonder if we're CPU bound here, since I do see CPU pegged at equiv of single core.

These are things you'd have to worry about if you actually cared about training a model on GPU, I don't think it's a serious concern for this example. (Obviously efficient training would be *nicer*, but personally it doesn't feel necessary—I'd rather an inefficient training example rather than none.)

Reply:
Yeah agreed, I think the most important part is that we show that you __can__ train a model on the GPU. I don't think that potential users will care much about the specific model